saleor/checkout/complete_checkout.py

```diff
@@ -764,96 +765,110 @@ def complete_checkout(
     """
     if site_settings is None:
         site_settings = Site.objects.get_current().settings
-
-    fetch_checkout_prices_if_expired(
-        checkout_info, manager, lines, discounts=discounts, site_settings=site_settings
+    checkout_for_update = (
+        Checkout.objects.select_for_update()
+        .filter(pk=checkout_info.checkout.pk)
+        .first()
     )
-
-    checkout = checkout_info.checkout
-    channel_slug = checkout_info.channel.slug
-    payment = checkout.get_last_active_payment()
-    try:
-        _prepare_checkout(
-            manager=manager,
-            checkout_info=checkout_info,
-            lines=lines,
-            discounts=discounts,
-            tracking_code=tracking_code,
-            redirect_url=redirect_url,
-            payment=payment,
-        )
-    except ValidationError as exc:
-        gateway.payment_refund_or_void(payment, manager, channel_slug=channel_slug)
-        raise exc
-
-    try:
-        order_data = _get_order_data(
-            manager, checkout_info, lines, discounts, site_settings
-        )
-    except ValidationError as exc:
-        gateway.payment_refund_or_void(payment, manager, channel_slug=channel_slug)
-        raise exc
-
-    customer_id = None
-    if payment and user:
-        customer_id = fetch_customer_id(user=user, gateway=payment.gateway)
-
-    action_required = False
-    action_data: Dict[str, str] = {}
-    if payment:
-        txn = _process_payment(
-            payment=payment,  # type: ignore
-            customer_id=customer_id,
-            store_source=store_source,
-            payment_data=payment_data,
-            order_data=order_data,
-            manager=manager,
-            channel_slug=channel_slug,
-        )
-
-        if txn.customer_id and user:
-            store_customer_id(user, payment.gateway, txn.customer_id)  # type: ignore
-
-        action_required = txn.action_required
-        if action_required:
-            action_data = txn.action_required_data
-            release_voucher_usage(
-                order_data.get("voucher"), order_data.get("user_email")
+    with transaction_with_commit_on_errors():
+        if checkout_for_update:
```

Comment:
If you have that many lines inside an if-condition or a with-block, I find it more readable if you inverted the condition and returned from the function early, then you wouldn't have to keep the deeper indent for so many lines. Example:

```python
checkout_for_update = Checkout.something.first()
if not checkout_for_update:
   return  # some meaningful return from the function here

with transaction():
    # the actual logic
```